### PR TITLE
strengthen requirements for CL_DEVICE_TYPE query

### PR DESCRIPTION
**Pull request by bashbaug**
Proposed fix for #224.

This PR:

1. Moves the requirement that `CL_DEVICE_TYPE_CUSTOM` must not be combined with any other device types to the description of `CL_DEVICE_TYPE`.  (see (4) below)
2. Adds the requirement that `CL_DEVICE_TYPE_DEFAULT` must only be used to get device IDs using `clGetDeviceIDs` and must NOT be returned as the `CL_DEVICE_TYPE` for any device.
3. Clarifies that `CL_DEVICE_TYPE_ALL` also must only be used to get device IDs using `clGetDeviceIDs`.
4. ~~Adds an informative note that most implementations should only return a single device type for `CL_DEVICE_TYPE`, and should not return a combination of device types.~~  Updated: Adds the requirement that the device must return only a single device type for `CL_DEVICE_TYPE`.

### CODE REVIEW

File: api/opencl_platform_layer.asciidoc [hunk 1]
```diff
@@ -239,34 +239,40 @@ include::{generated}/api/version-notes/CL_DEVICE_TYPE_ACCELERATOR.asciidoc[]
...
-For example, a {CL_DEVICE_TYPE_CPU} device may also be a
-{CL_DEVICE_TYPE_GPU} device, or a {CL_DEVICE_TYPE_ACCELERATOR} device
-may also be some other, more descriptive device type.
-{CL_DEVICE_TYPE_CUSTOM} devices must not be combined with any other
+[NOTE]
+====
+Although the standard allows an OpenCL device to report more than one device
+type in its {CL_DEVICE_TYPE}, many applications assume that devices will only
+report a single device type. Implementations are strongly encouraged to report a
+single device type for {CL_DEVICE_TYPE} and not to report a combination of
```
Comment (alycm) [inline api/opencl_platform_layer.asciidoc:273]: Text above says `{CL_DEVICE_TYPE_DEFAULT} **must not** be reported in {CL_DEVICE_TYPE}`, but the wording of this note is just **strongly encouraged**.

I'm not sure if any devices have reported more than one non-default device here, e.g. `GPU | ACCELERATOR`, which thing wording discourages but doesn't outright forbid.

**Review by alycm — APPROVED** [review]
Textually LGTM.  Group needs to approve overall change.

File: api/opencl_platform_layer.asciidoc (hunk 1 above)
Comment (bashbaug) [reply api/opencl_platform_layer.asciidoc]: I strengthened the requirement further and now the device must report only a single device type.

**Review by alycm — APPROVED** [review]
New wording still approved, it's very clear on returning a single value.

**Comment by bashbaug** [comment]
Had a thought: Would we want `CL_DEVICE_TYPE_DEFAULT` and `CL_DEVICE_TYPE_ALL` to be valid values to pass to `clCreateContextFromType`?  I think we would.

If so, I should adjust the descriptions of these devices types so it is clearer that this is allowed.  In other words, `CL_DEVICE_TYPE_ALL` could be passed to `clGetDeviceIDs` or to `clCreateContextFromType`, though it will never be returned as a device type for `CL_DEVICE_TYPE`.

**Comment by bashbaug** [comment]
I've updated this PR so it describes `clCreateContextFromType` also.  I think it's ready to go.

**Review by kpet — APPROVED** [review]

**Pull request merged**